### ISSUE TITLE
[SOLVED] Custom "orderby" by price in WooCommerce not working as expected

### ISSUE THREAD

**Issue by pedlima** [10up/ElasticPress#2703]
We are trying to create a custom order in shop page, which is perfectly working when we turn off EP, but not working at all with EP activated.

In Debug we cannot see the orderby changed like we see in other default WC orderby (such as price or alphabetic).

This is our basic function:

```
add_filter( 'woocommerce_get_catalog_ordering_args', 'maidot_custom_product_sorting' );
function maidot_custom_product_sorting( $args ) {
    if( isset( $_GET['orderby'] ) && 'unitpriceasc' === $_GET['orderby'] ) {
        $args['meta_key'] = '_unit_price';
        $args['orderby'] = array( 'meta_value' => 'ASC' );
    }
    return $args;
}
```

And this is what we have already tried, so far with no changes when we check Debug:

1. Changed "meta_value" to "meta._price.double"
2. We tried the one above both with and without the array
3. Tried to apply: $args['orderby'] = apply_filters( 'ep_set_default_sort', '_unit_price' );
4. Tried to use the action 'pre_get_posts' to change the orderby

None worked so far so we are stuck...

Any help will be much appreciated!

**Comment by felipeelia** [comment]
Hi @pedlima,

You'll need `pre_get_posts` for that but with a priority higher than 11, as the EP function will set that order in the query [here](https://github.com/10up/ElasticPress/blob/develop/includes/classes/Feature/WooCommerce/WooCommerce.php#L797) using 11. I've tested this code and it seems to have the effect you want:
```
add_action(
	'pre_get_posts',
	function ( $query ) {
		if ( ! $query->is_main_query() || is_admin() ) {
			return;
		}
		if ( isset( $_GET['orderby'] ) && 'unitpriceasc' === $_GET['orderby'] ) {
			$query->set( 'orderby', 'meta._unit_price.value' );
			$query->set( 'order', 'ASC' );
		}
	},
	12
);
```

An important note: As `_unit_price` starts with `_` it won't be indexed by default. You'll need a code like the following (and then run a sync) so Elasticsearch has this field:
```
add_filter(
	'ep_prepare_meta_allowed_protected_keys',
	function ( $allowed ) {
		$allowed[] = '_unit_price';
		return $allowed;
	}
);
```

**Comment by pedlima** [comment]
> Hi @pedlima,
> 
> You'll need `pre_get_posts` for that but with a priority higher than 11, as the EP function will set that order in the query [here](https://github.com/10up/ElasticPress/blob/develop/includes/classes/Feature/WooCommerce/WooCommerce.php#L797) using 11. I've tested this code and it seems to have the effect you want:
> 
> ```
> add_action(
> 	'pre_get_posts',
> 	function ( $query ) {
> 		if ( ! $query->is_main_query() || is_admin() ) {
> 			return;
> 		}
> 		if ( isset( $_GET['orderby'] ) && 'unitpriceasc' === $_GET['orderby'] ) {
> 			$query->set( 'orderby', 'meta._unit_price.value' );
> 			$query->set( 'order', 'ASC' );
> 		}
> 	},
> 	12
> );
> ```
> 
> An important note: As `_unit_price` starts with `_` it won't be indexed by default. You'll need a code like the following (and then run a sync) so Elasticsearch has this field:
> 
> ```
> add_filter(
> 	'ep_prepare_meta_allowed_protected_keys',
> 	function ( $allowed ) {
> 		$allowed[] = '_unit_price';
> 		return $allowed;
> 	}
> );
> ```

Thank you so much for your help but unfortunately, it doesn't seem to be working. :(

The results are exactly the same. In Debug args I see this:
["orderby"]=> string(21) "menu_order title date"

And also this:
["meta_key"]=> string(11) "_unit_price"

And this:
["order"]=> string(3) "ASC"

So in one hand it does seem to assume the meta_key and order (which is what happened in the past already) but, just as it was before, still not placing the correct "orderby" as it should... :(

**Comment by pedlima** [comment]
@felipeelia Can you please give me a hand in here? Thank you very much in advance!

**Comment by pedlima** [comment]
@felipeelia I'm sorry to insist, but I would really appreciate some help with this. If you want, I can open a new issue, but I think it makes no sense. Thank you very much in advance!

**Comment by felipeelia** [comment]
Hi @pedlima. I've checked locally again and I can definitely see ` 'orderby' => string 'meta._unit_price.value' (length=22)` here when accessing https://elasticpress.test/shop/?orderby=unitpriceasc.

The fact that it is not working for you might indicate something related specifically to your codebase, which would require some additional investigation. As I said in #2696, that would fall outside of our scope here but it's something we'd be happy to help you with through a consultation, just let us know.

Thanks!

**Comment by pedlima** [comment]
For future reference to other visitors, what ended up being the only problem here was that instead of "meta._unit_price.value" I should use "meta._unit_price.value.sortable" and with this, all worked flawlessly!

Now, this is once again very frustrating because your documentation is very incomplete. Maybe you just do it on purpose so that people can't have access to much information and have to hire your consultancy services, or maybe I am being unfair and it's only a matter of not having the necessary time to improve documentation.

Anyway, here's the solution for my case and I hope it helps future visitors. :)

Edit: Ok, just noticed that with this, it does not count with decimals. So independently of using comma or period, it always ignores and assumes the full value. So 17,38€ always appear before 2,38€ because it reads it like 1738 and 2380... it there a way to read decimals? With that, this would definitely be complete.

Edit 2: It's "meta._unit_price.double". Now it works counting with decimals. ;)

**Comment by brandwaffle** [comment]
@pedlima Our team has provided you with quite a bit of free help in this forum, including for the custom work you were looking to do on your site. Your response here is quite unwarranted and really seems to miss the point of how open source software works. I hope you're able to better understand this in the future and approach future support interactions with open source maintainers with some professionalism and patience. Best of luck with your use of ElasticPress.

**Comment by pedlima** [comment]
@brandwaffle I may have looked a bit "aggressive" in my comments, but still you do have a paid service for the plugin. It's not that you don't have that, so in this case, whoever is using your paid hosting and your plugin but still wants to check documentation on how to properly use it, I just shared my personal opinion regarding the lack of information on how your plugin works and how information in missing to whoever wants to dig further by their own.

Take it as a constructive criticism. I do know how open source works and if not, I would not take the time to come here only to tell the community how I solved my problem. I think you also was not correct in saying that without first looking into what I've done to help the community.

So in the end, the purpose is to have a better code and better documentation which makes it easier to understand for everyone.

There's always room for improvement and that's what a open source community is used for. :)
